pybamm/solvers/base_solver.py

```diff
@@ -49,6 +75,294 @@ def atol(self):
+        -------
+        y0_consistent : array-like, same shape as y0_guess
+            Initial conditions that are consistent with the algebraic equations (roots
+            of the algebraic equations)
+        """
+        pybamm.logger.info("Start calculating consistent initial conditions")
+        rhs = model.rhs_eval
+        algebraic = model.algebraic_eval
+        y0_guess = model.concatenated_initial_conditions.flatten()
+        jac = model.jac_algebraic_eval
+
+        # Split y0_guess into differential and algebraic
+        len_rhs = rhs(0, y0_guess).shape[0]
+        y0_diff, y0_alg_guess = np.split(y0_guess, [len_rhs])
+
+        def root_fun(y0_alg):
+            "Evaluates algebraic using y0_diff (fixed) and y0_alg (changed by algo)"
+            y0 = np.concatenate([y0_diff, y0_alg])
+            out = algebraic(0, y0)
+            pybamm.logger.debug(
+                "Evaluating algebraic equations at t=0, L2-norm is {}".format(
+                    np.linalg.norm(out)
+                )
+            )
+            return out
+
+        if jac:
+            if issparse(jac(0, y0_guess)):
+
+                def jac_fn(y0_alg):
+                    """
+                    Evaluates jacobian using y0_diff (fixed) and y0_alg (varying)
+                    """
+                    y0 = np.concatenate([y0_diff, y0_alg])
+                    return jac(0, y0)[:, len_rhs:].toarray()
+
+            else:
+
+                def jac_fn(y0_alg):
+                    """
+                    Evaluates jacobian using y0_diff (fixed) and y0_alg (varying)
+                    """
+                    y0 = np.concatenate([y0_diff, y0_alg])
+                    return jac(0, y0)[:, len_rhs:]
+
+        else:
+            jac_fn = None
+        # Find the values of y0_alg that are roots of the algebraic equations
+        sol = optimize.root(
+            root_fun,
+            y0_alg_guess,
+            jac=jac_fn,
+            method=self.root_method,
+            tol=self.root_tol,
+        )
+        # Return full set of consistent initial conditions (y0_diff unchanged)
+        y0_consistent = np.concatenate([y0_diff, sol.x])
+
+        if sol.success and np.all(sol.fun < self.root_tol * len(sol.x)):
+            pybamm.logger.info("Finish calculating consistent initial conditions")
+            return y0_consistent
+        elif not sol.success:
+            raise pybamm.SolverError(
+                "Could not find consistent initial conditions: {}".format(sol.message)
+            )
+        else:
+            raise pybamm.SolverError(
+                """
+                Could not find consistent initial conditions: solver terminated
+                successfully, but maximum solution error ({}) above tolerance ({})
+                """.format(
+                    np.max(sol.fun), self.root_tol * len(sol.x)
+                )
+            )
+
     def solve(self, model, t_eval, external_variables=None, inputs=None):
```

Comment:
Does external variables actually work when just solving? I thought it was only implemented for "stepping"

Reply:
Yes, it should work identically with both now - though I guess the main use cases will be for stepping